pkg/storage/backend-storage/backend-storage.go

```diff
@@ -107,6 +111,152 @@ func PVCForMigrationTarget(pvcStore cache.Store, migration *corev1.VirtualMachin
 	return pvcForMigrationTargetFromStore(pvcStore, migration)
 }
 
+func RecoverFromBrokenMigration(client kubecli.KubevirtClient, migration *corev1.VirtualMachineInstanceMigration, pvcStore cache.Store, vmi *corev1.VirtualMachineInstance, launcherImage string) error {
+	if migration.Status.MigrationState == nil ||
+		migration.Status.MigrationState.TargetPersistentStatePVCName == migration.Status.MigrationState.SourcePersistentStatePVCName {
+		// The migration either didn't actually start, or the backend storage is RWX.
+		// In both cases we consider the migration as failed.
+		migration.Status.Phase = corev1.MigrationFailed
+		return nil
+	}
+
+	// An interrupted migration exists. Using a job to check if the source PVC contains /meta/migrated,
+	// which would indicate that the libvirt migration finished.
+	// A JobComplete condition indicates the file is present, the migration was successful and the target PVC prevails
+	// A JobFailed condition indicated the file is absent, the migration didn't finish and the source PVC prevails
+
+	jobName := "recover-" + migration.Name
+
+	job, err := client.BatchV1().Jobs(vmi.Namespace).Get(context.Background(), jobName, metav1.GetOptions{})
+	if err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return err
+		}
+		job = buildRecoveryJob(jobName, migration.Status.MigrationState.SourcePersistentStatePVCName, launcherImage)
+		job, err = client.BatchV1().Jobs(vmi.Namespace).Create(context.Background(), job, metav1.CreateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+
+	// Give the job 5 seconds to complete
+	err = virtwait.PollImmediately(time.Second, 5*time.Second, func(ctx context.Context) (done bool, err error) {
```

Comment:
5 Seconds is still a lot  in my opinion. You may be recovering migration equal to #of workers but less than total cluster limit. As mentioned before I would make this async or move the recovery to its own controller.

Reply:
@xpivarc what objects do you think the external controller will operate on? 
I'm a bit worried, as I don't think we should finalize the migration until we identify that it is over.
Instead of waiting for the job to finish, we could raise a condition on the VMIM and run the job. The migration controller will not finalize the VMIM until the condition is removed.
The external controller could monitor the jobs and remove the condition from a relevant VMIM...
Can you think of a simpler flow?